### ISSUE TITLE
Implement modular framework for construction of decorated chemical environment proposals

### ISSUE THREAD

**Issue by davidlmobley** [open-forcefield-group/smarty#48]
This proposes a new framework for modular creation fo more complicated and sophisticated decorators ( introduced in #13, with a first pass  in #39). We think more modularization will be needed than we have now.

Here's a recap of where we were and are:
- Prior to #39, new atom types were proposed by `&`-ing existing types with decorators, which can get us **some** of what is in parm@frosst but not everything, i.e. issues as in this comment: https://github.com/open-forcefield-group/smarty/issues/13#issuecomment-227512655 
- In #39, we have a first pass which deals with this issue by applying potentially more than one decorator at once, potentially allowing discovery of more complex types like a hydrogen bonded to a carbon that has an electron withdrawing group attached to it directly from just knowing about hydrogen (which may be necessary in AlkEthOH since there is no score benefit to finding the `HC` subtype of hydrogen from the generic if one already has `HO`; one has to find `H1` before there is a score benefit). 

We're still looking at the possible solution in #39. It works by picking a base atom type to start with, and 1-3 decorators to use. These decorators are drawn from the updated set of decorators in https://github.com/open-forcefield-group/smarty/blob/binaries/examples/AlkEtOH2/atomtypes/decorators.smarts and include bond type information, connectivity, hydrogen count, formal charge, and aliphatic or aromatic. When the chosen decorators involve a bond, such as the below, additional base atom types are chosen to fill in where the `z` character is in the bond expression: 

```
$(*-z) singly bonded
$(*=z) doubly bonded
$(*#z) triply bonded
$(*:z) aromatic bond
$(*~z) any bond
```

So, we're creating patterns which can include up to three base atom types in their SMARTS patterns, which is good as it covers substantially more chemical space. 

**However**, we think that more modularization will be needed. To explain this modularization, let's first introduce the concept of alpha and beta substituents. Given a particular atom type we are considering, an alpha substituent is its first layer of neighbors. Beta substituents are the second layer of neighbors. SMARTS pattern `[#1]` has no alpha substituent and is only a single atom. SMARTS pattern `[#1$(*-[#6])]` has an alpha substituent. SMARTS pattern  `[#1$(*-[#6]-[#8])]` has an alpha substituent and a beta substituent. Multiple beta substituents are possible, as in `[#1$(*-[#6](-[#8])-[#8])]`.

After discussion with @cbayly13 and @bannanc , and discussion of the substituent issues above with @camizanette , we want to propose something like this framework for construction of decorated atom types: 
- `AtomDecorator( atomtype, decorators )` - take a specified atomtype apply one of the provided decorators to it, returning a tuple containing a SMARTS string corresponding to a decorated atomtype and the corresponding name
- `PickAnAtom( atomtypes)` - takes a desired current set of atomtypes (current, base, populated base, or some combination of the above or something else) and pick one at random.  (Here, we will **typically** but perhaps not always want to use **currently populated** atomtypes, i.e. the `atomtypes` list, not `basetypes` (generic types) or `used_basetypes` (base types which type anything)). A little silly, but still helpful. Returns a tuple corresponding to the selected (smarts, name).
- `HasAlpha(atomtype)` - takes a specified atomtype tuple (smarts, name) and determines whether or not it already has an alpha substituent, returning `True` or `False`.
- `AddAlphaSubstituentAtom( atom1type, bondset, atom2type)`: Takes specified atomtypes `atom1type` and `atom2type` (where `atom1type` is a type without an alpha substituent), and a specified bond set, and introduces an alpha substituent involving `atom2type` (which can be a decorated type such as output by `AtomDecorator`); returns a tuple of (smarts, name) from the result. Example output for input of `atom1type=("[#1]", "hydrogen")`, `bondset = (['-'], "single")`, and `atom2type = ("[#6]"` is `("[#1$(*-[#6])]", "hydrogen singly bonded to carbon"`) or something along those lines. This should basically just be adding `$(*zw)` where `z` is the selected bond type and `w` is `atom2type`. It should raise an exception if an alpha substituent is attempted to be added to an `atom1type` which already has an alpha substituent.
- `AddBetaSubstituentAtom( atom1type, bondset, atom2type)`: Takes specified atomtypes `atom1type` and `atom2type` (where `atom1type` is a type WITH an alpha substituent), and a specified bond set, and introduces an alpha substituent involving `atom2type` (which can be a decorated type such as output by `AtomDecorator`); returns a tuple of (smarts, name) from the result. Example output for input of `atom1type=("[#1$(*-[#6])]", "hydrogen singly bonded to carbon")`, `bondset = (['-'], "single")`, and `atom2type = ("[#8]"` is `("[#1$(*-[#6]-[#8])]", "hydrogen singly bonded to (carbon singly bonded to oxygen)"`) or something along those lines. This will have to handle two cases separately -- addition of the first beta substituent (where it is inserted between the characters `(]` and is inserted without parens), and addition of subsequent beta substituents (where it is inserted after the first set of `[]` in the substituent and is inserted enclosed in parens, i.e. `(-[w])`.)  It should raise an exception if a beta substituent is attempted to be added to an `atom1type` which does not have an alpha substituent.

`AddAlpha...` and `AddBeta...` are separated because we anticipate later reuse in different ways.

`BondToAtom` will be used, for example, to do things like:
`BondToAtom( bondset, atom1type, AtomDecorator( PickAnAtom(atomtypes), decorators) )`. But it can also be applied recursively, i.e. if we want to sometimes throw in moves which take us further down the tree at once, we can decorate an atom by decorating it not by bonding it to another atom but by bonding it to another GROUP which we compose on the fly, doing something like this:
`BondToAtom( bondset, atom1type, BondToAtom(bondset, AtomDecorator(PickAnAtom( atomlist), decorators), AtomDecorator(PickAnAtom(atomlist), decorators) )`

Substituent addition will be done in this type of framework:

``` Python
atom1type = PickAnAtom( populated_atomtypes )
hasalpha = HasAlpha(atom1type)
if hasalpha:
    newtype = AddAlphaSubstituent(atom1type, bondset, AtomDecorator(basetypes))
else:
    newtype = AddBetaSubstituent(atom1type, bondset, AtomDecorator(basetypes))
```

or something along those lines  

This should modularize quite a bit in a way that will allow us to extend more and try different things without having to wrangle regexes as much, etc., we think.

Ultimately, we imagine that `PickAnAtom` will be overloaded to allow constraints such as asking it for an atom which can only have one bond, or a valence 2 or higher atom, or one with available valence of at least X, etc., for efficiency reasons (avoiding generating things which make no chemical sense).

@camizanette plans to make the first attempt at this.

**Comment by bannanc** [comment]
> ("-",'simply")

This is a minor detail, but that should probably be called singly not simply.

**Comment by davidlmobley** [comment]
Fixing the above to correct that, thanks @bannanc .

**Comment by davidlmobley** [comment]
I'm making updates to the above after round 2 of our big confab.

**Comment by davidlmobley** [comment]
OK, I updated the above, @camizanette . @bannanc , maybe check for me to make sure I haven't botched anything.

**Comment by camizanette** [comment]
I am starting to work on this. Thanks!

**Comment by davidlmobley** [comment]
Resolved by #59 . Thanks, @camizanette !
